File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoSetStateInRender.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  CompilerDiagnostic,
  CompilerError,
  ErrorCategory,
} from '../CompilerError';
import {
  HIRFunction,
  Identifier,
  IdentifierId,
  Instruction,
  InstructionValue,
  isPrimitiveType,
  isSetStateType,
  Phi,
  Place,
  SpreadPattern,
} from '../HIR';
import {computeUnconditionalBlocks} from '../HIR/ComputeUnconditionalBlocks';
import {eachInstructionValueOperand} from '../HIR/visitors';
import {Result} from '../Utils/Result';

function isPrimitiveSetArg(
  arg: Place | SpreadPattern,
  fn: HIRFunction,
): boolean {
  if (arg.kind !== 'Identifier') {
    return false;
  }

  const visited = new Set<IdentifierId>();
  const defs = buildDefinitionMap(fn);
  return isPrimitiveIdentifier(arg.identifier, defs, visited);
}

type DefinitionMap = Map<IdentifierId, Instruction | Phi>;

function buildDefinitionMap(fn: HIRFunction): DefinitionMap {
  const defs: DefinitionMap = new Map();

  for (const [, block] of fn.body.blocks) {
    for (const phi of block.phis) {
      defs.set(phi.place.identifier.id, phi);
    }
    for (const instr of block.instructions) {
      defs.set(instr.lvalue.identifier.id, instr);
    }
  }

  return defs;
}

function isPrimitiveIdentifier(
  identifier: Identifier,
  defs: DefinitionMap,
  visited: Set<IdentifierId>,
): boolean {
  if (isPrimitiveType(identifier)) {
    return true;
  }

  if (visited.has(identifier.id)) {
    return false;
  }
  visited.add(identifier.id);

  const def = defs.get(identifier.id);
  if (def == null) {
    return false;
  }

  if (isPhi(def)) {
    return Array.from(def.operands.values()).every(operand =>
      isPrimitiveIdentifier(operand.identifier, defs, visited),
    );
  }

  return isPrimitiveInstruction(def.value, defs, visited);
}

function isPhi(def: Instruction | Phi): def is Phi {
  return 'kind' in def && def.kind === 'Phi';
}

function isPrimitiveInstruction(
  value: InstructionValue,
  defs: DefinitionMap,
  visited: Set<IdentifierId>,
): boolean {
  switch (value.kind) {
    case 'Primitive':
    case 'TemplateLiteral':
    case 'JSXText':
    case 'UnaryExpression':
    case 'BinaryExpression':
      return true;

    case 'TypeCastExpression':
      return isPrimitiveIdentifier(value.value.identifier, defs, visited);

    case 'LoadLocal':
    case 'LoadContext':
      return isPrimitiveIdentifier(value.place.identifier, defs, visited);

    case 'StoreLocal':
    case 'StoreContext':
      return isPrimitiveIdentifier(value.value.identifier, defs, visited);

    case 'Await':
      return isPrimitiveIdentifier(value.value.identifier, defs, visited);

    default:
      return false;
  }
}

/**
 * Validates that the given function does not have an infinite update loop
 * caused by unconditionally calling setState during render. This validation
 * is conservative and cannot catch all cases of unconditional setState in
 * render, but avoids false positives. Examples of cases that are caught:
 *
 * ```javascript
 * // Direct call of setState:
 * const [state, setState] = useState(false);
 * setState(true);
 *
 * // Indirect via a function:
 * const [state, setState] = useState(false);
 * const setTrue = () => setState(true);
 * setTrue();
 * ```
 *
 * However, storing setState inside another value and accessing it is not yet
 * validated:
 *
 * ```
 * // false negative, not detected but will cause an infinite render loop
 * const [state, setState] = useState(false);
 * const x = [setState];
 * const y = x.pop();
 * y();
 * ```
 */
export function validateNoSetStateInRender(
  fn: HIRFunction,
): Result<void, CompilerError> {
  const unconditionalSetStateFunctions: Set<IdentifierId> = new Set();
  return validateNoSetStateInRenderImpl(fn, unconditionalSetStateFunctions);
}

function validateNoSetStateInRenderImpl(
  fn: HIRFunction,
  unconditionalSetStateFunctions: Set<IdentifierId>,
): Result<void, CompilerError> {
  const unconditionalBlocks = computeUnconditionalBlocks(fn);
  const enableUseKeyedState = fn.env.config.enableUseKeyedState;
  let activeManualMemoId: number | null = null;
  const errors = new CompilerError();
  for (const [, block] of fn.body.blocks) {
    for (const instr of block.instructions) {
      switch (instr.value.kind) {
        case 'LoadLocal': {
          if (
            unconditionalSetStateFunctions.has(instr.value.place.identifier.id)
          ) {
            unconditionalSetStateFunctions.add(instr.lvalue.identifier.id);
          }
          break;
        }
        case 'StoreLocal': {
          if (
            unconditionalSetStateFunctions.has(instr.value.value.identifier.id)
          ) {
            unconditionalSetStateFunctions.add(
              instr.value.lvalue.place.identifier.id,
            );
            unconditionalSetStateFunctions.add(instr.lvalue.identifier.id);
          }
          break;
        }
        case 'ObjectMethod':
        case 'FunctionExpression': {
          if (
            // faster-path to check if the function expression references a setState
            [...eachInstructionValueOperand(instr.value)].some(
              operand =>
                isSetStateType(operand.identifier) ||
                unconditionalSetStateFunctions.has(operand.identifier.id),
            ) &&
            // if yes, does it unconditonally call it?
            validateNoSetStateInRenderImpl(
              instr.value.loweredFunc.func,
              unconditionalSetStateFunctions,
            ).isErr()
          ) {
            // This function expression unconditionally calls a setState
            unconditionalSetStateFunctions.add(instr.lvalue.identifier.id);
          }
          break;
        }
        case 'StartMemoize': {
          CompilerError.invariant(activeManualMemoId === null, {
            reason: 'Unexpected nested StartMemoize instructions',
            description: null,
            details: [
              {
                kind: 'error',
                loc: instr.value.loc,
                message: null,
              },
            ],
          });
          activeManualMemoId = instr.value.manualMemoId;
          break;
        }
        case 'FinishMemoize': {
          CompilerError.invariant(
            activeManualMemoId === instr.value.manualMemoId,
            {
              reason:
                'Expected FinishMemoize to align with previous StartMemoize instruction',
              description: null,
              details: [
                {
                  kind: 'error',
                  loc: instr.value.loc,
                  message: null,
                },
              ],
            },
          );
          activeManualMemoId = null;
          break;
        }
        case 'CallExpression': {
          const callee = instr.value.callee;
          if (
            isSetStateType(callee.identifier) ||
            unconditionalSetStateFunctions.has(callee.identifier.id)
          ) {
            if (activeManualMemoId !== null) {
              errors.pushDiagnostic(
                CompilerDiagnostic.create({
                  category: ErrorCategory.RenderSetState,
                  reason:
                    'Calling setState from useMemo may trigger an infinite loop',
                  description:
                    'Each time the memo callback is evaluated it will change state. This can cause a memoization dependency to change, running the memo function again and causing an infinite loop. Instead of setting state in useMemo(), prefer deriving the value during render. (https://react.dev/reference/react/useState)',
                  suggestions: null,
                }).withDetails({
                  kind: 'error',
                  loc: callee.loc,
                  message: 'Found setState() within useMemo()',
                }),
              );
            } else if (unconditionalBlocks.has(block.id)) {
              let isArgPrimitive = false;

              if (instr.value.args.length > 0) {
                const arg = instr.value.args[0];
                if (arg.kind === 'Identifier') {
                  isArgPrimitive = isPrimitiveSetArg(arg, fn);
                }
              }

              if (isArgPrimitive) {
                if (enableUseKeyedState) {
                  errors.pushDiagnostic(
                    CompilerDiagnostic.create({
                      category: ErrorCategory.RenderSetState,
                      reason:
                        'Calling setState during render may trigger an infinite loop',
                      description:
                        'Use useKeyedState instead of calling setState directly in render. Example: const [value, setValue] = useKeyedState(initialValue, key)',
                      suggestions: null,
                    }).withDetails({
                      kind: 'error',
                      loc: callee.loc,
                      message: 'Found setState() in render',
                    }),
                  );
                }
              } else {
                errors.pushDiagnostic(
                  CompilerDiagnostic.create({
                    category: ErrorCategory.RenderSetState,
                    reason:
                      'Calling setState during render may trigger an infinite loop',
                    description:
                      'Calling setState during render will trigger another render, and can lead to infinite loops. (https://react.dev/reference/react/useState)',
                    suggestions: null,
                  }).withDetails({
                    kind: 'error',
                    loc: callee.loc,
                    message: 'Found setState() in render',
                  }),
                );
              }
            }
          }
          break;
        }
      }
    }
  }

  return errors.asResult();
}

File: compiler/crates/react_compiler_validation/src/validate_no_set_state_in_render.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! Validates that the function does not unconditionally call setState during render.
//!
//! Port of ValidateNoSetStateInRender.ts.

use std::collections::HashSet;

use react_compiler_diagnostics::{
    CompilerDiagnostic, CompilerDiagnosticDetail, ErrorCategory,
};
use react_compiler_hir::dominator::compute_unconditional_blocks;
use react_compiler_hir::environment::Environment;
use react_compiler_hir::{
    BlockId, HirFunction, Identifier, IdentifierId, InstructionValue, PlaceOrSpread, Type,
};

pub fn validate_no_set_state_in_render(func: &HirFunction, env: &mut Environment) {
    let mut unconditional_set_state_functions: HashSet<IdentifierId> = HashSet::new();
    let next_block_id = env.next_block_id().0;
    let diagnostics = validate_impl(
        func,
        &env.identifiers,
        &env.types,
        &env.functions,
        next_block_id,
        env.config.enable_use_keyed_state,
        &mut unconditional_set_state_functions,
    );
    for diag in diagnostics {
        env.record_diagnostic(diag);
    }
}

fn is_set_state_id(
    identifier_id: IdentifierId,
    identifiers: &[Identifier],
    types: &[Type],
) -> bool {
    let ident = &identifiers[identifier_id.0 as usize];
    let ty = &types[ident.type_.0 as usize];
    react_compiler_hir::is_set_state_type(ty)
}

fn validate_impl(
    func: &HirFunction,
    identifiers: &[Identifier],
    types: &[Type],
    functions: &[HirFunction],
    next_block_id_counter: u32,
    enable_use_keyed_state: bool,
    unconditional_set_state_functions: &mut HashSet<IdentifierId>,
) -> Vec<CompilerDiagnostic> {
    let unconditional_blocks: HashSet<BlockId> =
        compute_unconditional_blocks(func, next_block_id_counter);
    let mut active_manual_memo_id: Option<u32> = None;
    let mut errors: Vec<CompilerDiagnostic> = Vec::new();

    for (_block_id, block) in &func.body.blocks {
        for &instr_id in &block.instructions {
            let instr = &func.instructions[instr_id.0 as usize];
            match &instr.value {
                InstructionValue::LoadLocal { place, .. } => {
                    if unconditional_set_state_functions.contains(&place.identifier) {
                        unconditional_set_state_functions.insert(instr.lvalue.identifier);
                    }
                }
                InstructionValue::StoreLocal { lvalue, value, .. } => {
                    if unconditional_set_state_functions.contains(&value.identifier) {
                        unconditional_set_state_functions
                            .insert(lvalue.place.identifier);
                        unconditional_set_state_functions
                            .insert(instr.lvalue.identifier);
                    }
                }
                InstructionValue::ObjectMethod { lowered_func, .. }
                | InstructionValue::FunctionExpression { lowered_func, .. } => {
                    let inner_func = &functions[lowered_func.func.0 as usize];

                    // Check if any operand references a setState
                    // For function expressions, the operands are the context captures
                    // plus any explicit operands in the instruction value
                    let has_set_state_operand = {
                        // Check context variables
                        let mut found = inner_func.context.iter().any(|ctx_place| {
                            is_set_state_id(ctx_place.identifier, identifiers, types)
                                || unconditional_set_state_functions
                                    .contains(&ctx_place.identifier)
                        });
                        if !found {
                            // Also check the instruction value operands (dependencies)
                            // In TS: eachInstructionValueOperand checks deps for FunctionExpression
                            found = inner_func.context.iter().any(|ctx_place| {
                                unconditional_set_state_functions
                                    .contains(&ctx_place.identifier)
                            });
                        }
                        found
                    };

                    if has_set_state_operand {
                        let inner_errors = validate_impl(
                            inner_func,
                            identifiers,
                            types,
                            functions,
                            next_block_id_counter,
                            enable_use_keyed_state,
                            unconditional_set_state_functions,
                        );
                        if !inner_errors.is_empty() {
                            unconditional_set_state_functions
                                .insert(instr.lvalue.identifier);
                        }
                    }
                }
                InstructionValue::StartMemoize {
                    manual_memo_id, ..
                } => {
                    active_manual_memo_id = Some(*manual_memo_id);
                }
                InstructionValue::FinishMemoize {
                    manual_memo_id, ..
                } => {
                    active_manual_memo_id = None;
                    let _ = manual_memo_id;
                }
                InstructionValue::CallExpression { callee, .. } => {
                    if is_set_state_id(callee.identifier, identifiers, types)
                        || unconditional_set_state_functions
                            .contains(&callee.identifier)
                    {
                        if active_manual_memo_id.is_some() {
                            errors.push(
                                CompilerDiagnostic::new(
                                    ErrorCategory::RenderSetState,
                                    "Calling setState from useMemo may trigger an infinite loop",
                                    Some(
                                        "Each time the memo callback is evaluated it will change state. This can cause a memoization dependency to change, running the memo function again and causing an infinite loop. Instead of setting state in useMemo(), prefer deriving the value during render. (https://react.dev/reference/react/useState)".to_string(),
                                    ),
                                )
                                .with_detail(CompilerDiagnosticDetail::Error {
                                    loc: callee.loc,
                                    message: Some("Found setState() within useMemo()".to_string()),
                                }),
                            );
                        } else if unconditional_blocks.contains(&block.id) {
                            if enable_use_keyed_state {
                                errors.push(
                                    CompilerDiagnostic::new(
                                        ErrorCategory::RenderSetState,
                                        "Cannot call setState during render",
                                        Some(
                                            "Calling setState during render may trigger an infinite loop.\n\
                                            * To reset state when other state/props change, use `const [state, setState] = useKeyedState(initialState, key)` to reset `state` when `key` changes.\n\
                                            * To derive data from other state/props, compute the derived data during render without using state".to_string(),
                                        ),
                                    )
                                    .with_detail(CompilerDiagnosticDetail::Error {
                                        loc: callee.loc,
                                        message: Some("Found setState() in render".to_string()),
                                    }),
                                );
                            } else {
                                errors.push(
                                    CompilerDiagnostic::new(
                                        ErrorCategory::RenderSetState,
                                        "Cannot call setState during render",
                                        Some(
                                            "Calling setState during render may trigger an infinite loop.\n\
                                            * To reset state when other state/props change, store the previous value in state and update conditionally: https://react.dev/reference/react/useState#storing-information-from-previous-renders\n\
                                            * To derive data from other state/props, compute the derived data during render without using state".to_string(),
                                        ),
                                    )
                                    .with_detail(CompilerDiagnosticDetail::Error {
                                        loc: callee.loc,
                                        message: Some("Found setState() in render".to_string()),
                                    }),
                                );
                            }
                        }
                    }
                }
                _ => {}
            }
        }
    }

    errors
}
